### PR TITLE
make sure userdata is correctly passed

### PR DESCRIPTION
**Pull request by DaanHoogland**
## Description

removing class="disallowSpecialCharacters" to prevent url encoding
and calling sanitizeReverse on the fields contents before encoding it to send to the server.

## Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

## How Has This Been Tested?

This has been manually tested by adding user data containing the characters '&', '<' and '>' in the simulator UI and then checking the data base field in the user_vm table to see if the characters where xml-entities after base64-decoding them.

### CODE REVIEW

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
doing it here because of local db problems :(

File: ui/index.html [hunk 1]
```diff
@@ -557,7 +557,7 @@ <h3><translate key="label.select.iso.or.template" /></h3>
                                             <span><translate key="label.add.userdata"/> (<translate key="label.optional"/>)</span>
                                         </div>
                                         <div class="value">
-                                            <textarea name="userdata" class="disallowSpecialCharacters"></textarea>
+                                            <textarea name="userdata"></textarea>
```
Comment (DaanHoogland) [inline ui/index.html:560]: this might not be needed but it makes no sense either, so I'll keep this in.

File: ui/scripts/instanceWizard.js [hunk 2]
```diff
@@ -1094,8 +1094,15 @@
...
+
+                // NOTE code duplicated from sanatizeReverse in ui/utils.js because of call propagation issues
+                userdata = userdata
+                            .replace(/&amp;/g, "&")
+                            .replace(/&lt;/g, "<")
+                            .replace(/&gt;/g, ">");
+
                 $.extend(deployVmData, {
-                    userdata : encodeURIComponent(btoa(userdata))
+                    userdata : btoa(userdata)
```
Comment (DaanHoogland) [inline ui/scripts/instanceWizard.js]: I'll re-add the encodeURI call as a base64 string may contain '=' signs.

File: ui/scripts/instanceWizard.js [hunk 3]
```diff
@@ -1094,8 +1094,15 @@
 
             var userdata = args.data.userdata;
             if (userdata != null && userdata.length > 0) {
+
+                // NOTE code duplicated from sanatizeReverse in ui/utils.js because of call propagation issues
+                userdata = userdata
+                            .replace(/&amp;/g, "&")
+                            .replace(/&lt;/g, "<")
+                            .replace(/&gt;/g, ">");
```
Comment (DaanHoogland) [inline ui/scripts/instanceWizard.js]: Nicolas reported he had a test case where the calling variity did work, I so I'll experiment more to see if I can get it to work and replace above with a call to cloudstack.sanitizeReverse(userdata);

**Review by yadvr — APPROVED** [review]
LGTM, if testing is okay @DaanHoogland then send to apache/cloudstack

**Review by nvazquez — APPROVED** [review]

**Comment by DaanHoogland** [comment]
moved to master

**Comment by DaanHoogland** [comment]
see https://github.com/apache/cloudstack/pull/3260

**Pull request closed without merging**